ARROW-11477: [R][Doc] Reorganize and improve README and vignette content (#10014)
apache/arrow: r/README.md

[![CI](https://github.com/apache/arrow/workflows/R/badge.svg?event=push)](https://github.com/apache/arrow/actions?query=workflow%3AR+branch%3Amaster+event%3Apush)
[![conda-forge](https://img.shields.io/conda/vn/conda-forge/r-arrow.svg)](https://anaconda.org/conda-forge/r-arrow)

**[Apache Arrow](https://arrow.apache.org/) is a cross-language
development platform for in-memory data.** It specifies a standardized
language-independent columnar memory format for flat and hierarchical
data, organized for efficient analytic operations on modern hardware. It
also provides computational libraries and zero-copy streaming messaging
and interprocess communication.

**The `arrow` package exposes an interface to the Arrow C++ library,
enabling access to many of its features in R.** It provides low-level
access to the Arrow C++ library API and higher-level access through a
`dplyr` backend and familiar R functions.

## What can the `arrow` package do?

-   Read and write **Parquet files** (`read_parquet()`,
    `write_parquet()`), an efficient and widely used columnar format
-   Read and write **Feather files** (`read_feather()`,
    `write_feather()`), a format optimized for speed and

>     interoperability

Author: Agreed—I cut the words here down to the bone to keep each bullet point on one line when rendered in GitHub

>     interoperability

Author: I added some more detail on Parquet further down. I think for now it would be best not to add more links or detail in this section.

-   Open or write **large, multi-file datasets** with a single function
    call (`open_dataset()`, `write_dataset()`)

> -   Read **large CSV and JSON files** with excellent **speed and

Author: I don't really think this is the right place for links; this is mean to be a fairly breezy list of key features, with elaboration provided below. I'd prefer to table this idea for later

    efficiency** (`read_csv_arrow()`, `read_json_arrow()`)
-   Read and write files in **Amazon S3** buckets with no additional
    function calls
-   Exercise **full control over data types** of columns when reading
    and writing data files
-   Use **compression codecs** including Snappy, gzip, Brotli,
    Zstandard, LZ4, LZO, and bzip2 for reading and writing data
-   Manipulate and analyze **larger-than-memory datasets** with
    **`dplyr` verbs**
-   Pass data between **R and Python** in the same process
-   Connect to **Arrow Flight** RPC servers to send and receive large
    datasets over networks
-   Access and manipulate Arrow objects through **low-level bindings**
    to the C++ library
-   Provide a **toolkit for building connectors** to other applications

>     and services that use Arrow

Author: This **making words bold** thing is kinda cheesy, I know, but it stops this from just looking like a big wall of text.

## Installation

Install the latest release of `arrow` from CRAN with

``` r
install.packages("arrow")
```

Conda users can install `arrow` from conda-forge with

``` shell
conda install -c conda-forge --strict-channel-priority r-arrow
```

Installing a released version of the `arrow` package requires no
additional system dependencies. For macOS and Windows, CRAN hosts binary
packages that contain the Arrow C++ library. On Linux, source package
installation will also build necessary C++ dependencies. For a faster,
more complete installation, set the environment variable
`NOT_CRAN=true`. See `vignette("install", package = "arrow")` for
details.

> ## Installing a development version

Reviewer: Maybe delete this heading?

Author: I think it's worthwhile to set this apart from the release install info. I'll try changing this and the other to subheadings under an "Installation" heading

`help("list_compute_functions")`.

For `dplyr` queries on `Table` objects, if the `arrow` package detects
an unimplemented function within a `dplyr` verb, it automatically calls
`collect()` to return the data as an R `data.frame` before processing
that `dplyr` verb. For queries on `Dataset` objects (which can be larger
than memory), it raises an error if the function is unimplemented.

### Other uses

Other uses of `arrow` are described in the following vignettes:

-   `vignette("python", package = "arrow")`: use `arrow` and
    `reticulate` to pass data between R and Python
-   `vignette("flight", package = "arrow")`: connect to Arrow Flight RPC
    servers to send and receive data
-   `vignette("arrow", package = "arrow")`: access and manipulate Arrow
    objects through low-level bindings to the C++ library

## Getting help

If you encounter a bug, please file an issue with a minimal reproducible
example on the [Apache Jira issue
tracker](https://issues.apache.org/jira/projects/ARROW/issues). Create
an account or log in, then click **Create** to file an issue. Select the
project **Apache Arrow (ARROW)**, select the component **R**, and begin

> the issue summary with **\[R\]** followed by a space. For more

Reviewer 2: On my rendering, the escape of `\[` isn't actually working.

I can't quite tell from these issues if there's a better work around or if this is resolved in the development version of pkgdown (they have switched from markdown_github to gfm) 

https://github.com/rstudio/rmarkdown/issues/1860
https://github.com/r-lib/pkgdown/pull/1193
https://github.com/r-lib/pkgdown/pull/1195

Author: Huh, weird, thanks. Easy dodge: I'll just make it `monospace` so no escapes are needed.

Reviewer 2: Yeah, I tried the "obvious" `\\[` `\\\[` and nothing worked easily without monospacing it

[![CI](https://github.com/apache/arrow/workflows/R/badge.svg?event=push)](https://github.com/apache/arrow/actions?query=workflow%3AR+branch%3Amaster+event%3Apush)
[![conda-forge](https://img.shields.io/conda/vn/conda-forge/r-arrow.svg)](https://anaconda.org/conda-forge/r-arrow)

**[Apache Arrow](https://arrow.apache.org/) is a cross-language
development platform for in-memory data.** It specifies a standardized
language-independent columnar memory format for flat and hierarchical
data, organized for efficient analytic operations on modern hardware. It
also provides computational libraries and zero-copy streaming messaging
and interprocess communication.

**The `arrow` package exposes an interface to the Arrow C++ library,
enabling access to many of its features in R.** It provides low-level
access to the Arrow C++ library API and higher-level access through a
`dplyr` backend and familiar R functions.

> ## What can the `arrow` package do?

Reviewer: ```suggestion
What can the `arrow` package do?
```

Author: I think it looks much better with this set as a heading

packages that contain the Arrow C++ library. On Linux, source package
installation will also build necessary C++ dependencies. For a faster,
more complete installation, set the environment variable
`NOT_CRAN=true`. See `vignette("install", package = "arrow")` for
details.

### Installing a development version

Development versions of the package (binary and source) are built
nightly and hosted at <https://arrow-r-nightly.s3.amazonaws.com>. To
install from there:

``` r
install.packages("arrow", repos = "https://arrow-r-nightly.s3.amazonaws.com")
```

Conda users can install `arrow` nightly builds with

``` shell
conda install -c arrow-nightlies -c conda-forge --strict-channel-priority r-arrow
```

If you already have a version of `arrow` installed, you can switch to
the latest nightly development version with

``` r
arrow::install_arrow(nightly = TRUE)
```

These nightly package builds are not official Apache releases and are
not recommended for production use. They may be useful for testing bug
fixes and new features under active development.

## Usage

Among the many uses of the `arrow` package, two of the most accessible
uses are:

> -   Reading and writing data files

Reviewer: So what? Isn't it pretty basic and universal to be able to read and write files?

I know you're trying to be concise, but maybe it would be good to note reasons you'd want to use `arrow` to read your files, such as (1) supports different file formats; (2) fast; (3) can read and write to cloud storage.

> -   Manipulating Arrow data with `dplyr` verbs

Reviewer: ```suggestion
-   Manipulating bigger-than-memory data with `dplyr` verbs
```

    afford better interoperability with databases and data warehouse
    systems
-   `Dataset`: a data structure functionally similar to `Table` but with
    the capability to work on larger-than-memory data partitioned across
    multiple files

### Reading and writing data files with `arrow`

The `arrow` package provides functions for reading single data files in
several common formats. By default, calling any of these functions
returns an R `data.frame`. To return an Arrow `Table`, set argument
`as_data_frame = FALSE`.

-   `read_parquet()`: read a file in Parquet format
-   `read_feather()`: read a file in Feather format (the Apache Arrow
    IPC format)
-   `read_delim_arrow()`: read a delimited text file (default delimiter
    is comma)
-   `read_csv_arrow()`: read a comma-separated values (CSV) file
-   `read_tsv_arrow()`: read a tab-separated values (TSV) file
-   `read_json_arrow()`: read a JSON data file

For writing data to single files, the `arrow` package provides the
functions `write_parquet()` and `write_feather()`. These can be used
with R `data.frame` and Arrow `Table` objects.

For example, let’s write the Star Wars characters data that’s included

> in `dplyr` to a Parquet file, then read it back in. First load the

Reviewer: Why would I want to do this? Perhaps here is where you put the note/link about the Parquet format.

`arrow` and `dplyr` packages:

``` r
library(arrow, warn.conflicts = FALSE)
library(dplyr, warn.conflicts = FALSE)
```

Then write the `data.frame` named `starwars` to a Parquet file at
`file_path`:

``` r
file_path <- tempfile()
write_parquet(starwars, file_path)
```

Then read the Parquet file into an R `data.frame` named `sw`:

``` r
sw <- read_parquet(file_path)

> ```

Reviewer: It would also be good here to mention (and link to the "arrow" vignette) that R attributes are preserved round-trip, so you can save things like `sf` data frames to Parquet. That seems to be a use case of interest.

For reading and writing larger files or multiple files with Arrow

> `Dataset` objects, `arrow` provides the functions `open_dataset()` and

Reviewer: I would mention here that this does not require reading all of the data into memory, so you can filter and select data for analysis efficiently

> `write_dataset()`. For examples of these, see

Reviewer: Writing datasets is not just about the inverse of reading, it's about partitioning, which lets you speed up your queries significantly

`vignette("dataset", package = "arrow")`.

All these functions can read and write files in the local filesystem (by

> passing unqualified paths or `file://` URIs) or files in Amazon S3 (by

Reviewer: I wouldn't mention `file://` URIs, I think it's confusing

passing S3 URIs beginning with `s3://`). For more details, see
`vignette("fs", package = "arrow")`

### Using `dplyr` with `arrow`

The `arrow` package provides a `dplyr` backend enabling manipulation of
Arrow tabular data with `dplyr` verbs. To use it, first load both
packages `arrow` and `dplyr`. Then load data into an Arrow `Table` or
`Dataset` object. For example, read the Parquet file written in the
previous example into an Arrow `Table` named `sw`:

``` r
sw <- read_parquet(file_path, as_data_frame = FALSE)
```

Next, pipe on `dplyr` verbs:

``` r
result <- sw %>% 
  filter(homeworld == "Tatooine") %>% 
  rename(height_cm = height, mass_kg = mass) %>%
  mutate(height_in = height_cm / 2.54, mass_lbs = mass_kg * 2.2046) %>%
  arrange(desc(birth_year)) %>%
  select(name, height_in, mass_lbs)
```

> The `arrow` package uses lazy evaluation to delay computation until the

Reviewer: Why is delaying computation useful?

#> height_in: expr
#> mass_lbs: expr
#> 
#> * Filter: equal(homeworld, "Tatooine")
#> * Sorted by birth_year [desc]
#> See $.data for the source Arrow object
```

To execute these computations and materialize the result, call
`compute()` or `collect()`. `compute()` returns an Arrow `Table`,
suitable for passing to other `arrow` or `dplyr` functions:

``` r
result %>% compute()
#> Table
#> 10 rows x 3 columns
#> $name <string>
#> $height_in <double>
#> $mass_lbs <double>
```

`collect()` returns an R `data.frame`, suitable for viewing or passing
to other R functions for analysis or visualization:

``` r
result %>% collect()
#> # A tibble: 10 x 3
#>    name               height_in mass_lbs
#>    <chr>                  <dbl>    <dbl>
#>  1 C-3PO                   65.7    165. 
#>  2 Cliegg Lars             72.0     NA  
#>  3 Shmi Skywalker          64.2     NA  
#>  4 Owen Lars               70.1    265. 
#>  5 Beru Whitesun lars      65.0    165. 
#>  6 Darth Vader             79.5    300. 
#>  7 Anakin Skywalker        74.0    185. 
#>  8 Biggs Darklighter       72.0    185. 
#>  9 Luke Skywalker          67.7    170. 
#> 10 R5-D4                   38.2     70.5
```

> The `arrow` package works with most `dplyr` verbs except those that

Reviewer: This is the part where I think that a full dplyr vignette would help, so you could get into these details and provide more examples. I think the support we have is greater than this paragraph implies (not that the words are wrong, it just focuses on what "limited" support there is rather than enumerating all of the things that are supported).

The NEWS.md now has some more discussion that you could include here.

compute aggregates (such as `summarise()`, and `mutate()` after
`group_by()`). Inside `dplyr` verbs, Arrow offers limited support for
functions and operators, with broader support expected in upcoming
releases. For more information about available compute functions, see

> `help("list_compute_functions")`.

Reviewer: I don't think we want to link to this on the README. This should be deep in the (hypothetical) vignette. I would expect that most users won't call the arrow_namespaced compute functions but will just use the base/stringr/dplyr-mapped versions. Unfortunately there's not a good place to list those (though you could envision adding a function that could list them), which is why I was thinking a vignette would be necessary to document them.

For writing data to single files, the `arrow` package provides the
functions `write_parquet()` and `write_feather()`. These can be used
with R `data.frame` and Arrow `Table` objects.

For example, let’s write the Star Wars characters data that’s included
in `dplyr` to a Parquet file, then read it back in. First load the
`arrow` and `dplyr` packages:

``` r
library(arrow, warn.conflicts = FALSE)
library(dplyr, warn.conflicts = FALSE)
```

Then write the `data.frame` named `starwars` to a Parquet file at
`file_path`:

``` r
file_path <- tempfile()
write_parquet(starwars, file_path)
```

Then read the Parquet file into an R `data.frame` named `sw`:

``` r
sw <- read_parquet(file_path)
```

For reading and writing larger files or multiple files with Arrow
`Dataset` objects, `arrow` provides the functions `open_dataset()` and
`write_dataset()`. For examples of these, see
`vignette("dataset", package = "arrow")`.

All these functions can read and write files in the local filesystem (by
passing unqualified paths or `file://` URIs) or files in Amazon S3 (by
passing S3 URIs beginning with `s3://`). For more details, see
`vignette("fs", package = "arrow")`

> ### Using dplyr with Arrow

Author: We discussed the possibility of adding a separate new vignette about using arrow with dplyr. When I started to do that, I found that most of what needed to be said was not about the dplyr backend functionality but about the Arrow objects that you call them on and the Arrow objects they return. So the dplyr vignette was really turning out to be less than half about dplyr, and mostly about the data reading and writing functions and about `Table` and `Dataset` objects. Since that content belongs here in the README, I think it makes sense to slip in the dplyr content here too, at least for now, especially since the new dplyr capabilities are a marquee feature for the release. But if there are strong feelings that the README is not the right place for these 55 lines of dplyr content, I'm happy to discuss alternatives.

| Dim | Class          | Description                             | How to create an instance                                     |
|-----|----------------|-----------------------------------------|---------------------------------------------------------------|
| 0   | `Scalar`       | single value and its `DataType`         | `Scalar$create(value, type)`                                  |
| 1   | `Array`        | vector of values and its `DataType`     | `Array$create(vector, type)`                                  |
| 1   | `ChunkedArray` | vectors of values and their `DataType`  | `ChunkedArray$create(..., type)`                              |
| 2   | `RecordBatch`  | list of `Array`s with a `Schema`        | `RecordBatch$create(...)`                                     |
| 2   | `Table`        | list of `ChunkedArray` with a `Schema`  | `Table$create(...)` or `arrow::read_*(as_data_frame = FALSE)` |
| 2   | `Dataset`      | list of `Table`s with the same `Schema` | see `vignette("dataset", package = "arrow")`                  |

Each of these is defined as an `R6` class in the `arrow` R package and
corresponds to a class of the same name in the Arrow C++ library. The

> `arrow` package provides a variety of `R6` and S3 methods for

Reviewer 2: Very minor
 
```suggestion
`arrow` package provides a variety of `R6` and `S3` methods for
```

Author: Elsewhere in our docs, S3 is not set in monospace text, but `R6` is. Was that deliberate?

interacting with instances of these classes.

## Reading and writing data files with Arrow

The `arrow` package provides functions for reading data from several
common file formats. By default, calling any of these functions returns
an R data frame. To return an Arrow `Table`, set argument

> `as_data_frame = FALSE`.

Reviewer 2: Would it be too much of a distraction to have a brief aside about why one might want to do this here?

Author: I think the "Why" section I'm working on adding at the top of the README should answer that 👍

-   `read_parquet()`: read a file in Parquet format (an efficient
    columnar data format)
-   `read_feather()`: read a file in Feather format (the Apache Arrow
    IPC format)
-   `read_delim_arrow()`: read a delimited text file (default delimiter
    is comma)
-   `read_csv_arrow()`: read a comma-separated values (CSV) file
-   `read_tsv_arrow()`: read a tab-separated values (TSV) file
-   `read_json_arrow()`: read a JSON data file

For writing Arrow tabular data structures to files, the `arrow` package
provides the functions `write_parquet()` and `write_feather()`. These

> functions also accept R data frames.

Reviewer 2: ```suggestion
functions also accept R dataframes.
```

Author: In R, it should be either "data frame" or "data.frame"—don't get me started on how it's totally different if we're talking about pandas or Spark...

Reviewer 2: I would go for `data.frame` personally, "data frame" looks super weird and 1990s to me!

## Using dplyr with Arrow

The `arrow` package provides a `dplyr` backend, enabling manipulation of
Arrow tabular data with `dplyr` verbs. To begin, load both `arrow` and
`dplyr`:

``` r
library(arrow, warn.conflicts = FALSE)
library(dplyr, warn.conflicts = FALSE)
```

Then create an Arrow `Table` or `RecordBatch` using one of the object
creation or file loading functions listed above. For example, create a
`Table` named `sw` with the Star Wars characters data frame that’s
included in `dplyr`:

``` r
sw <- Table$create(starwars)
```

Or read the same data from a Parquet file, using `as_data_frame = FALSE`
to create a `Table` named `sw`:

``` r
write_parquet(starwars, data_file <- tempfile()) # write file to demonstrate reading it
sw <- read_parquet(data_file, as_data_frame = FALSE)
```

For larger or multi-file datasets, load the data into a `Dataset` as

> described in `vignette("dataset", package = "arrow")`.

Reviewer 2: I wonder if this would be better below the `dplyr` block below? It might be the GH diff UI, but I thought this was semi-distracting from the example we're working with so far here

Author: I think it's worth having here (or possibly in both places) but I'll add an "Or, " at the beginning to make it clearer that it's just another data loading option like the other two listed immediately above.

Next, pipe on `dplyr` verbs:

``` r
result <- sw %>% 
  filter(homeworld == "Tatooine") %>% 
  rename(height_cm = height, mass_kg = mass) %>%
  mutate(height_in = height_cm / 2.54, mass_lbs = mass_kg * 2.2046) %>%
  arrange(desc(birth_year)) %>%
  select(name, height_in, mass_lbs)
```

> The `arrow` package uses lazy evaluation. `result` is an object with

Reviewer 2: ```suggestion
The `arrow` package uses lazy evaluation to delay computation until it is requested. `result` is an object with
```

Author: I think I'll use "until the result is required" instead of "until if that is requested" if that's cool with you

Reviewer 2: Yeah, that works — I had that at first but then went down a path of "but what does 'required' mean and how do we operationalize it?" but I think that's because I have a developer's perspective

``` r
result %>% collect()
#> # A tibble: 10 x 3
#>    name               height_in mass_lbs
#>    <chr>                  <dbl>    <dbl>
#>  1 C-3PO                   65.7    165. 
#>  2 Cliegg Lars             72.0     NA  
#>  3 Shmi Skywalker          64.2     NA  
#>  4 Owen Lars               70.1    265. 
#>  5 Beru Whitesun lars      65.0    165. 
#>  6 Darth Vader             79.5    300. 
#>  7 Anakin Skywalker        74.0    185. 
#>  8 Biggs Darklighter       72.0    185. 
#>  9 Luke Skywalker          67.7    170. 
#> 10 R5-D4                   38.2     70.5
```

Arrow supports most `dplyr` verbs except those that compute aggregates
(such as `summarise()`, and `mutate()` after `group_by()`). Inside
`dplyr` verbs, Arrow offers limited support for functions and operators,
with broader support expected in upcoming releases. For more information
about available compute functions, see `help("list_compute_functions")`.

For `dplyr` queries on `Table` and `RecordBatch` objects, if the `arrow`
package detects an unsupported function within a `dplyr` verb, it
automatically calls `collect()` to return the data as an R data frame

> before processing that `dplyr` verb.

Reviewer 2: Should we add something here about what happens on unsupported calls with datasets?

## Getting help

If you encounter a bug, please file an issue with a minimal reproducible
example on [the Apache Jira issue
tracker](https://issues.apache.org/jira/). Choose the project **Apache
Arrow (ARROW)**, select the component **R**, and begin the issue summary
with **\[R\]** followed by a space. For more information, see the
**Report bugs and propose features** section of the [Contributing to
Apache
Arrow](https://arrow.apache.org/docs/developers/contributing.html) page
in the Arrow developer documentation.

> We welcome questions and discussion about the `arrow` package. For

Reviewer 2: ```suggestion
We welcome questions and discussion about and contributions to the `arrow` package. For
```

(or something similar)

Install the latest release of `arrow` from CRAN with

``` r
install.packages("arrow")
```

Conda users can install `arrow` from conda-forge with

    conda install -c conda-forge --strict-channel-priority r-arrow

Installing a released version of the `arrow` package requires no
additional system dependencies. For macOS and Windows, CRAN hosts binary
packages that contain the Arrow C++ library. On Linux, source package
installation will also build necessary C++ dependencies. For a faster,
more complete installation, set the environment variable
`NOT_CRAN=true`. See `vignette("install", package = "arrow")` for
details.

## Installing a development version

Development versions of the package (binary and source) are built
nightly and hosted at <https://arrow-r-nightly.s3.amazonaws.com>. To
install from there:

``` r
install.packages("arrow", repos = "https://arrow-r-nightly.s3.amazonaws.com")
```

> Or to switch to the latest nightly development version:

Reviewer: ```suggestion
If you have `arrow` installed and want to switch to the latest nightly development version, you can use the included `install_arrow()` utility function:
```

``` r
arrow::install_arrow(nightly = TRUE)
```

> Conda users can install `arrow` nightly builds with

Author: To clarify: do you mean I should move
``` shell
conda install -c arrow-nightlies -c conda-forge --strict-channel-priority r-arrow
```
to above
``` r
arrow::install_arrow(nightly = TRUE)
```
?

> Conda users can install `arrow` nightly builds with

Reviewer: correct. install_arrow is a convenience for both the CRAN and conda methods of installation

    conda install -c arrow-nightlies -c conda-forge --strict-channel-priority r-arrow

These nightly package builds are not official Apache releases and are
not recommended for production use. They may be useful for testing bug
fixes and new features under active development.

## Apache Arrow metadata and data objects

> Arrow defines the following classes for representing metadata:

Author: The benefit of keeping it in this order is that there are no forward references; reading from top to bottom, you never come across something that has not already been defined above.

| Class      | Description                                      | How to create an instance        |
|------------|--------------------------------------------------|----------------------------------|
| `DataType` | attribute controlling how values are represented | functions in `help("data-type")` |
| `Field`    | string name and a `DataType`                     | `field(name, type)`              |
| `Schema`   | list of `Field`s                                 | `schema(...)`                    |

Arrow defines the following classes for representing 0-dimensional
(scalar), 1-dimensional (vector), and 2-dimensional (tabular/data

> frame-like) data:

Reviewer: I might set this up differently: start with what R has since we can assume users are familiar with that. R has vectors, and data.frame that collects vectors of equal length. Arrow has two in-memory representations for the same data: one, RecordBatch containing Arrays, is closest to the R model. The other, Table containing ChunkedArrays, supports array data that may be sliced into multiple pieces in order to accommodate larger sizes or other optimizations. A Dataset is similar to a Table but does not hold all of the data in memory.

Then get into differences. Arrow has a true Scalar type, though you as an R user won't generally have to deal with them. Arrow has a richer data type system and metadata (discuss Schema, DataType, and the range of data types there are.)

Author: Ok, gotcha. Based on this and your other comment above, I think it makes sense to add a "Why use this package?" section before the "Installation" section that will establish the framing to put the details here in more of a clearer context.

I could also modify the content here to focus chiefly on the `Table` class and move these tables into a separate vignette, perhaps titled "Apache Arrow object hierarchy" (because it's valuable to have this content presented _somewhere_ in a documentation style as opposed to a narrative style, even if it's not here in the readme).

Author: I moved the object hierarchy content to the "Using the Arrow C++ Library in R" vignette and added just two bullets here describing `Table` and `Dataset` which for many purposes will be the only two data structures users really need to know about.

| Dim | Class          | Description                               | How to create an instance                          |
|-----|----------------|-------------------------------------------|----------------------------------------------------|
| 0   | `Scalar`       | single value and its `DataType`           | `Scalar$create(value, type)`                       |
| 1   | `Array`        | vector of values and its `DataType`       | `Array$create(vector, type)`                       |

> | 1   | `ChunkedArray` | list of `Array`s with the same `DataType` | `chunked_array(..., type)`                         |

Reviewer: For standardization maybe this should use `ChunkedArray$create()`, just so the pattern is clear. Also IDK that "list of `Array`s with the same `DataType`", while accurate on some level, is the right thing to say here. In terms of interface, it behaves like Array, you don't treat it like a list type.

| 2   | `RecordBatch`  | list of `Array`s with a `Schema`          | `record_batch(...)`                                |

> | 2   | `Table`        | list of `ChunkedArray` with a `Schema`    | `Table$create(...)` or `arrow::read_*()` functions |

Reviewer: perhaps?

```suggestion
| 2   | `Table`        | list of `ChunkedArray` with a `Schema`    | `Table$create(...)` or `arrow::read_*(as_data_frame = FALSE)` functions |
```

| 2   | `Dataset`      | list of `Table`s with the same `Schema`   | see `vignette("dataset", package = "arrow")`       |

These classes exist in the `arrow` R package and correspond to classes
of the same names in the Arrow C++ library. For convenience, the `arrow`
package also defines several synthetic classes that do not exist in the
C++ library, including:

> -   `ArrowDatum`: inherited by `Scalar`, `Array`, and `ChunkedArray`

Reviewer: I wouldn't include this in the README. This is an implementation detail. There are implementation details in vignette("arrow").

-   `ArrowTabular`: inherited by `RecordBatch` and `Table`
-   `ArrowObject`: inherited by all Arrow objects

These are all defined as `R6` classes. The `arrow` package provides a
variety of `R6` and S3 methods for interacting with instances of these
classes.

## Reading and writing data files with Arrow

The `arrow` package provides functions for reading data from several
common file formats. By default, calling any of these functions returns
an R data frame. To return an Arrow `Table`, set argument
`as_data_frame = FALSE`.

-   `read_delim_arrow()`: read a delimited text file (default delimiter
    is comma)
-   `read_csv_arrow()`: read a comma-separated values (CSV) file
-   `read_tsv_arrow()`: a tab-separated values (TSV) file
-   `read_json_arrow()`: read a JSON data file

> -   `read_feather()`: read a file in Feather format (the Apache Arrow

Reviewer: I'd lead with parquet and feather

#>  1 C-3PO                   65.7    165. 
#>  2 Cliegg Lars             72.0     NA  
#>  3 Shmi Skywalker          64.2     NA  
#>  4 Owen Lars               70.1    265. 
#>  5 Beru Whitesun lars      65.0    165. 
#>  6 Darth Vader             79.5    300. 
#>  7 Anakin Skywalker        74.0    185. 
#>  8 Biggs Darklighter       72.0    185. 
#>  9 Luke Skywalker          67.7    170. 
#> 10 R5-D4                   38.2     70.5
```

Arrow supports most dplyr verbs except those that compute aggregates
(such as `summarise()` and `mutate()` after `group_by()`). Inside dplyr
verbs, Arrow offers limited support for functions and operators, with
broader support expected in upcoming releases. For more information
about available compute functions, see `help("list_compute_functions")`.

For dplyr queries on `Table` and `RecordBatch` objects, if the `arrow` R
package detects an unsupported function within a dplyr verb, it
automatically calls `collect()` to return the data as an R data frame
before processing that dplyr verb.

## Getting help

If you encounter a bug, please file an issue with a minimal reproducible
example on [the Apache Jira issue

> tracker](https://issues.apache.org/jira/). Choose the project **Apache

Reviewer: I think there is a direct link to the arrow JIRA, should be on our website on the contributing page.

Author: I considered this and tried first to use one of these URLs:
- https://issues.apache.org/jira/projects/ARROW/issues/
- https://issues.apache.org/jira/projects/ARROW/summary
but I tried opening these in incognito windows to see how users would see them if they're not already signed into Jira, and they were all quite confusing.

I also tried constructing a direct link to the Create Issue page by using URL parameters after https://issues.apache.org/jira/secure/CreateIssue.jspa but I couldn't get that working either

Author: I guess I'll just switch this to using https://​issues.apache.org/​jira/​projects/​ARROW/​issues because that's used everywhere else, but oof it's a bad experience for folks who aren't familiar with Jira
